### PR TITLE
feat(autorun): playbook-level HITL gate markers (#232)

### LINKED ISSUE #232: Experience Report: HITL Gates for Auto-Run Mode
## Context

Ran a 24-hour parallel session with two agents in separate git worktrees, each using a SpecFlow playbook. The workflow has natural review points where human judgment is needed before proceeding. Currently, auto-run skips past these because there's no pause-notify-resume mechanism.

## The Problem

In **interactive mode**, review gates happen naturally (you're watching). In **auto-run**, they get skipped. The agents completed full cycles without human input:

- Specs were created → agents immediately proceeded to implementation
- Plans were drafted → agents immediately started coding
- No opportunity for "wait, that's not quite right" before significant work was done

## Natural Review Points (Skipped in Auto-Run)

- After spec is written → Human should review spec before implementation begins
- After architecture is planned → Human should approve approach
- After implementation → Human should review code before verification
- After tests pass → Human should validate before marking complete

### Option A: Playbook-Level HITL Markers

Playbook documents could declare HITL gates:

```markdown
## Step 5: Review Specification

<!-- MAESTRO:HITL reason="Spec ready for review" artifact=".maestro/outputs/SPEC.md" -->

- [ ] Human has reviewed and approved the specification
```

Maestro pauses at the marker, sends notification with reason/artifact, waits for approval.

### Option B: Agent-to-Maestro Signaling

The agent (or underlying tool like SpecFlow) could signal Maestro:

```bash
maestro signal --pause --reason "Spec ready for review" --artifact "SPEC.md"
```

Or via file convention that Maestro monitors.

This would allow integration between workflow tools (like SpecFlow) and Maestro's control loop.

## Related

- Issue #231 - Exit condition detection (same session)
- Issue #201 - Observability during Auto Run

### PR DESCRIPTION
**Pull request by chr1syy**
Closes #232

## Summary

Adds support for playbook-level **HITL (human-in-the-loop) gates** in Auto Run — Option A from the issue. Authors mark review points in playbook documents with:

```markdown
## Step 5: Review Specification

<!-- MAESTRO:HITL reason="Spec ready for review" artifact=".maestro/outputs/SPEC.md" -->

- [ ] Human has reviewed and approved the specification
```

When Auto Run reaches a marker that is positioned before an unchecked task (with no intervening checked task), it pauses, surfaces a sticky toast with the reason + artifact, and the user gets the existing `AutoRunErrorBanner` with **Resume** and **Abort Run** buttons. Once the user ticks the human-approval checkbox, the marker is treated as consumed and the run continues to the next gate or end-of-doc.

This eliminates the orchestration burden that split playbooks (\`SpecFlow_1_Specify\` → \`SpecFlow_2_Plan\` → \`SpecFlow_3_Implement\`) currently put on users — the workflow now manages itself between phases, while the human stays in the driver's seat at decision points.

## Why

From the issue: in interactive mode, review gates happen naturally because you're watching. In Auto Run they get skipped, so 24-hour parallel sessions blow past review points and the agent has already done significant downstream work by the time you notice. Maestro should pause at natural review checkpoints (after spec, after plan, after impl, after tests) and only continue once a human approves.

Per the maintainer's question — _\"why not split playbooks into phases?\"_ — the reporter's response was that split playbooks invert the relationship: the human ends up managing sequencing, context handoffs, and \"which playbook is next?\" instead of just making approval decisions. Integrated gates keep one workflow, one context, with human-as-decider rather than human-as-orchestrator.

## How it works

The parser \`findPendingHitlGate(content)\` in \`batchUtils.ts\` walks the document line-by-line:

- Skips fenced code blocks (so docs can include the marker syntax as examples)
- Treats checked tasks as \"consuming\" any pending marker above them — the user already approved
- Returns the **first** marker in a pending chain when the next unchecked task is reached
- Returns null when an unchecked task appears before any marker (no gate above it)

In \`useBatchRunner.ts\`, the runner calls the parser at the top of the inner task loop (after the existing error-resolution \`await\` block, before \`processTask\`). On hit it synthesizes a recoverable \`AgentError { type: 'hitl_gate' }\` and routes through the existing \`pauseBatchOnError\` — so the **AutoRunErrorBanner** UI, the **PAUSED_ERROR** state machine entry, the \`errorResolutionRefs\` promise chain, and the Resume/Skip/Abort actions all just work without new UI plumbing. The toast uses \`dismissible: true\` so the user must click it (review gates shouldn't auto-dismiss).

After **Resume**, the loop re-reads the doc so the next HITL check sees fresh content — if the user ticked the box, the marker is now consumed and we continue; if they didn't, we re-pause.

## Changes

- \`src/shared/types.ts\` — add \`'hitl_gate'\` to the \`AgentErrorType\` union (1 line).
- \`src/renderer/hooks/batch/batchUtils.ts\` — add \`HitlGate\` type and \`findPendingHitlGate(content)\` parser. Reuses the existing fence-aware iteration pattern from \`countMarkdownTasks\`.
- \`src/renderer/hooks/batch/internal/useBatchRunner.ts\` — call the parser between iterations, synthesize the \`AgentError\`, route through \`pauseBatchOnError\`, emit a sticky toast. Re-read doc on resume.
- \`src/renderer/hooks/batch/useBatchProcessor.ts\` — thread \`pauseBatchOnError\` into the runner's deps.
- \`src/__tests__/renderer/hooks/batch/batchUtilsHitl.test.ts\` — 13 parser tests.

No new UI components. No new IPC. No main-process changes.

## Out of scope (deferred)

- **Option B** from the issue (CLI-based \`maestro signal --pause\` or file-watching). Option A covers the SpecFlow case cleanly; we can revisit if external workflow tools need it.
- Dedicated HITL banner styling (currently reuses the orange/red error banner). The text says \"Auto Run Paused\" + the reason, which reads correctly for HITL, but a dedicated \"Review requested\" theme could land in a follow-up.
- Marker validation/lint inside the playbook editor.
- Telemetry for HITL gate usage.

## Parser test coverage

- no marker present
- no tasks at all
- marker precedes the first unchecked task ✅
- unchecked task appears before any marker
- checked task consumes the marker above it
- re-pause on a fresh marker after a previously consumed one
- multiple markers in a chain → returns the first
- missing artifact attribute
- missing reason attribute (falls back to default)
- markers inside fenced code blocks are ignored
- real marker outside a fenced block is detected when an example exists inside one
- records 0-indexed line number
- CRLF line endings

## Test plan

- [x] \`npm run lint\` — TypeScript across all configs
- [x] \`npm run lint:eslint\` — ESLint on the changed files
- [x] \`npx prettier --check\` — all changed files
- [x] \`npx vitest run src/__tests__/renderer/hooks/batch/ src/__tests__/cli/services/batch-processor.test.ts\` — 213/213 passing, including 13 new HITL parser tests
- [ ] Manual: write a playbook with a HITL marker, start Auto Run, verify it pauses with the AutoRunErrorBanner and sticky toast, tick the human-approval box, click Resume, verify it continues to the next gate
- [ ] Manual: verify Abort Run from the HITL paused state cleanly stops the run
- [ ] Manual: verify markers inside fenced code blocks in the playbook docs don't trigger pauses

## Summary by CodeRabbit

* **New Features**
  * Human-in-the-loop (HITL) review gates: Auto-run pauses at HITL markers, shows resume/abort controls and a sticky info toast; pauses are deduplicated per marker and resume re-checks task state so only relevant tasks continue.

* **Tests**
  * Added test coverage validating HITL marker detection and handling across edge cases (consumed vs. pending markers, code blocks, line endings).

[Review Change Stack](https://app.coderabbit.ai/change-stack/RunMaestro/Maestro/pull/1015?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/renderer/hooks/batch/internal/useBatchRunner.ts [hunk 1]
```diff
@@ -624,6 +632,66 @@ export function useBatchRunner({
...
+							notifyToast({
+								type: 'info',
+								title: 'Auto Run paused for review',
+								message: hitlGate.artifact
+									? `${docEntry.filename}: ${hitlGate.reason} — review ${hitlGate.artifact}`
+									: `${docEntry.filename}: ${hitlGate.reason}`,
+								project: session.name,
+								sessionId,
+								dismissible: true,
+							});
```
[the remark this answers is not in the document]
Comment (chr1syy) [reply src/renderer/hooks/batch/internal/useBatchRunner.ts]: Fixed in `f63f37161` — added a per-run `activeHitlGateLine` local that tracks the line of the currently-paused gate. The toast and `autorun(...)` log fire only when `isNewGate` (activeHitlGateLine !== hitlGate.line). On every Resume-without-tick the parser returns the same line, `isNewGate` is false, and we skip the toast. The state is cleared when a different gate fires (multi-gate scenarios still get one toast each) or when no gate is detected after a task completes.

File: src/renderer/hooks/batch/internal/useBatchRunner.ts [hunk 2]
```diff
@@ -624,6 +632,66 @@ export function useBatchRunner({
...
+							// hand the agent a stale view of the playbook.
+							const {
+								taskCount: resumedRemaining,
+								checkedCount: resumedChecked,
+								content: resumedContent,
+							} = await readDocAndCountTasks(folderPath, effectiveFilename, sshRemoteId);
+							remainingTasks = resumedRemaining;
+							docCheckedCount = resumedChecked;
+							docContent = resumedContent;
+							if (remainingTasks === 0) break;
```
[the remark this answers is not in the document]
Comment (chr1syy) [reply src/renderer/hooks/batch/internal/useBatchRunner.ts]: Fixed in `f63f37161` — the re-read is now gated on `activeHitlGateLine !== null`, so non-HITL resume paths (agent_crashed, permission_denied, …) keep their pre-PR behavior of continuing with in-memory `docContent`. The HITL-scoped re-read is wrapped in try/catch: if the file was deleted or the SSH session dropped, we log a warn via `logger.warn` and continue with the in-memory content instead of crashing the runner.

**Comment by chr1syy** [comment]
Pushed `f63f37161` addressing both Greptile P1 comments. Threads replied inline.

**On the failing `test` job:** the failures are all in `src/__tests__/main/agents/claude-usage-startup.test.ts`, which doesn't exist on this branch — that file was added by #1010 (Maestro-P) merging to `rc` after this PR was opened. The same test job is failing on `rc` itself at `85b078177`:

```
https://github.com/RunMaestro/Maestro/actions/runs/26053802690
```

CI on `rc` passed at `b96e9a07e` (the commit this PR branched off) and started failing at the next merge. This is unrelated to the HITL changes here and should be fixed in `rc`. Happy to rebase once that's green, or leave the PR as-is.

**Comment by pedramamini** [comment]
Thanks for the contribution, @chr1syy — this is a really nicely scoped PR. 🙌

Reviewed the changes and the response to Greptile's two P1s, and I'm happy with it:

- The `activeHitlGateLine` dedupe is the right call — both for the toast-stacking issue and for keeping the doc re-read HITL-scoped so other recoverable-error resume paths keep their pre-PR behavior. The try/catch around `readDocAndCountTasks` is the right level of paranoia for a paused runner where the SSH session or file may have changed underneath us.
- Parser is fence-aware, handles CRLF, falls back gracefully on missing attributes, and reuses the existing `countMarkdownTasks` iteration pattern. 13 parser tests cover the edge cases I'd worry about (multi-marker chains, consumed-then-fresh, fenced examples with a real marker outside the fence).
- Reusing `pauseBatchOnError` + `AutoRunErrorBanner` instead of building new UI plumbing keeps the surface area tiny — exactly the kind of integration I'd want here.
- The big block comments above the HITL-resume re-read and the gate-detection block earn their weight by explaining *why* (re-pause-forever risk, dedupe rationale) rather than just narrating the code.

You're correct that the failing `test` job is unrelated — `claude-usage-startup.test.ts` was added on `rc` after this PR was opened (in #1010) and the same job is failing directly on `rc` at `85b078177`. No need to rebase on my account; I'll get `rc` green separately.

Labeling `approved` — go ahead and merge once you're ready.
